# Application of Bootstrap samples in Random Forest

In [0]:
import numpy as np
from sklearn.datasets import load_boston
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeRegressor

 <li> Load the boston house dataset </li>

In [0]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

#Concatenating arrays
data = np.hstack((boston.data, np.atleast_2d(boston.target).T))

In [0]:
x.shape

(506, 13)

In [0]:
data.shape

(506, 14)

In [0]:
# a = [[1,2,3,4,50],[5,6,7,8,100],[9,10,11,12,101],[13,14,15,16,102],[17,18,19,20,103]]
a = list(data)
import random, math
generated_samples_input = []
generated_samples_target = []
sample_models = []
selected_columns_30 = []
no_of_columns = [ random.randint(3,len(a[0])-1) for _ in range(30) ] #Generating list of no. of columns for each sample
for i in range(30):
  a60 = random.sample(a, k = math.floor(len(a)*0.6))
  a40 = random.sample(a60, k = math.floor(len(a) - len(a60)))
  final_sample = a60 + a40
  final_sample_array = np.array([np.array(xi) for xi in final_sample]) #Converting list of lists to array
  # print(final_sample_array)
  final_sample_y = final_sample_array[:, -1]
  sample_x = np.delete(final_sample_array, len(final_sample_array[0])-1, axis = 1)
  #Generating list of selected columns
  z = [i for i in range(0, len(a[0])-1)] #Column index generation
  selected_columns = random.sample(z,no_of_columns[i])
  selected_columns.sort()
  selected_columns_30.append(selected_columns)

  #Selecting columns from sample
  final_sample_x = sample_x[:, selected_columns]

  #Building model
  regressor = DecisionTreeRegressor(random_state=0).fit(final_sample_x, final_sample_y)
  sample_models.append(regressor)


  generated_samples_input.append(final_sample_x)
  generated_samples_target.append(final_sample_y)


In [0]:
# for input, target in zip(generated_samples_input, generated_samples_target):
#   for i,j in zip(input, target):
#     for model in sample_models:
#       model.predict(i.reshape(-1, 1))
#     break;


In [0]:
prediction = []
pred = 0
for input, model in zip(generated_samples_input, sample_models):
  pred += model.predict(input)
print(pred/30)

[24.78166667 24.58733333 23.24933333 25.03185185 20.26238095 24.92174129
 23.93533333 23.01007463 20.7248524  21.08340796 21.61578947 26.7345641
 22.71333333 22.85666667 23.64166667 21.48618574 22.13722222 20.43840796
 21.812      23.34866667 21.62611111 24.37388889 21.40507463 21.32666667
 23.42933333 23.27174129 19.046      21.03583333 23.38396352 24.93951907
 23.63571429 22.04416667 24.45507463 23.99714286 22.96174129 23.04174129
 19.19951907 23.5        25.75666667 20.42257143 24.66589744 23.24571429
 21.631      24.72578947 24.73194444 21.81007463 21.34666667 20.74674129
 22.22507463 20.57333333 20.27840796 27.39555556 26.05174129 21.04840796
 24.28666667 22.76222222 23.07166667 20.55507463 19.43888889 21.92255556
 23.31638889 20.65174129 22.016      22.46285714 22.11507463 21.98857143
 23.43396352 22.55844444 24.75644444 20.81285714 23.32377778 23.22
 24.09840796 20.58333333 26.30377778 22.09507463 17.57507463 23.842
 20.29533333 19.64333333 21.0765032  23.40174129 21.58333333 23

In [0]:
mse = round(mean_squared_error(y, pred/30), 6)
print(mse)

86.084753


In [0]:
# a = list(x)
# allsamples = []

# for i in range(30):
#   a60 = random.sample(a,k = math.floor(len(a)*0.6))
#   a40 = random.sample(a60,k = math.floor(len(a)*0.4))
#   final_sample = a60 + a40
#   print(final_sample)
#   for sample in final_sample:
#     columns = [ random.randint(3,x.shape[1]) for _ in range(30) ]
#     print(sample[:, columns])
#     break;
#   fs = random.sample(final_sample,columns[i])
#   allsamples.append(fs)

# # print(*allsamples[0], sep = '\n')

### Task: 1
<font color='red'><b>Step 1 Creating samples: </b></font> Randomly create 30 samples from the whole boston data points.
<ol>
<li>Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points</li>
<li>Ex: For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly consider we have selected [4, 5, 7, 8, 9, 3] now we will replciate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]</li>
<li> we create 30 samples like this </li>
<li> Note that as a part of the Bagging when you are taking the random samples make sure each of the sample will have                different set of columns</li>
<li> Ex: assume we have 10 columns for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample [7, 9, 1, 4, 5, 6, 2] and so on...</li>
<li> Make sure each sample will have atleast 3 feautres/columns/attributes</li>
</ol>

<font color='red'><b>Step 2 Building High Variance Models on each of the sample and finding train MSE value:</b></font> Build a DecisionTreeRegressor on each of the sample.
<ol><li>Build a regression trees on each of 30 samples.</li>
<li>computed the predicted values of each data point(506 data points) in your corpus.</li>
<li> predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.</li>
<li>Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.</li>
</ol>

<font color='red'><b>Step 3 Calculating the OOB score :</b></font>
<ol>
<li>Computed the predicted values of each data point(506 data points) in your corpus.</li>
<li>Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.</li>
<li>Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.</li>
</ol>

### Task: 2
<pre>
<font color='red'><b>Computing CI of OOB Score and Train MSE</b></font>
<ol>
<li> Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>
</pre>
### Task: 3
<pre>
<font color='red'><b>Given a single query point predict the price of house.</b></font>

<li>Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] Predict the house price for this point as mentioned in the step 2 of Task 1. </li>
</pre>